algorithms/czech.sbl

```diff
@@ -0,0 +1,115 @@
+routines (
+  RV R1
+  palatalise
+  mark_regions
+  do_possessive
+  do_case
+)
+
+externals ( stem )
+
+integers ( pV p1 )
+
+groupings ( v )
+
+stringescapes {}
+
+stringdef a' '{U+00E1}'
+stringdef c^ '{U+010D}'
+stringdef d^ '{U+010F}'
+stringdef e' '{U+00E9}'
+stringdef e^ '{U+011B}'
+stringdef i' '{U+00ED}'
+stringdef n^ '{U+0148}'
+stringdef o' '{U+00F3}'
+stringdef r^ '{U+0159}'
+stringdef s^ '{U+0161}'
+stringdef t^ '{U+0165}'
+stringdef u' '{U+00FA}'
+stringdef u* '{U+016F}'
+stringdef y' '{U+00FD}'
+stringdef z^ '{U+017E}'
+
+define v 'aeiouy{a'}{e^}{e'}{i'}{o'}{u'}{u*}{y'}'
+
+define mark_regions as (
+
+    $pV = limit
+    $p1 = limit
+
+    do (
+        gopast non-v setmark pV
+        gopast non-v gopast v setmark p1
+    )
```

Comment:
This isn't the usual Snowball definition of R1 (https://snowballstem.org/texts/r1r2.html) - I tried changing this to keep RV as above but make R1 match the usual definition:

```
    do ( gopast non-v setmark pV )
    do ( gopast v gopast non-v setmark p1 )
```

And I made a file with all the example Czech words mentioned in the paper, plus examples of all the noun and adjective declinations: [czech-words.txt](https://github.com/snowballstem/snowball/files/7103584/czech-words.txt)

Comparing the output of `./stemwords -l cs < czech-words.txt` using this custom R1 definition vs the standard R1 definition it seems pretty clear the standard R1 does a much better job for at least these cases.

@jimregan Can you remember the origin of this?

Reply:
No, I can't remember.

Reply:
I think I may have worked out why - consider e.g. `vlna` (wool) which has declension `vlna`, `vlny`, `vlne`, `vlnu`, `vlne`, `vlnou` - the aim would be to conflate these to the same stem (likely `vln`).  The normal Snowball R1 definition relies on vowels to measure the stem which doesn't work for these consonant clusters.

https://en.wikipedia.org/wiki/Czech_language#Consonants says:

> The consonants /r/, /l/, and /m/ can be [syllabic](https://en.wikipedia.org/wiki/Syllabic_consonant), acting as [syllable nuclei](https://en.wikipedia.org/wiki/Syllable_nuclei) in place of a vowel. [Strč prst skrz krk](https://en.wikipedia.org/wiki/Str%C4%8D_prst_skrz_krk) ("Stick [your] finger through [your] throat") is a well-known Czech [tongue twister](https://en.wikipedia.org/wiki/Tongue_twister) using syllabic consonants but no vowels.

And https://en.wikipedia.org/wiki/Czech_phonology#Consonant_chart says:

> Sonorants /r/, /l/ become [syllabic](https://en.wikipedia.org/wiki/Syllabic_consonant) between two consonants or after a consonant at the end of a word.

and https://en.wikipedia.org/wiki/Czech_phonology#Phonotactics:

> The syllabic nucleus is usually formed by vowels or diphthongs, but in some cases syllabic sonorants (/r/ and /l/, rarely also /m/ and /n/) can be found in the nucleus, e.g. vlk [vl̩k] ('wolf'), krk [kr̩k] ('neck'), osm [osm̩] ('eight').

So I wonder if the R1 definition for Czech needs to consider `r` and `l` (and perhaps `m` and `n`?) preceded by a consonant and not followed by a vowel as effectively implying a vowel.